dhall/src/Dhall/Import.hs

```diff
@@ -463,88 +465,376 @@ localToPath prefix file_ = liftIO $ do
+--   `processImportWithSemanticCache` in general.
+processImport
+  :: ImportStack -> ResolvedImport -> StateT Status IO SemanticImport
+processImport stack import_ = do
+    cache_ <- use cache
+    case Map.lookup import_ cache_ of
+        Just semanticImport -> return semanticImport
         Nothing -> do
-            exprFromUncachedImport here
-
-{-| Save an expression to the specified `Import`
-
-    Currently this only works for cached imports and ignores other types of
-    imports, but could conceivably work for uncached imports in the future
-
-    The main reason for this more general type is for symmetry with
-    `exprFromImport` and to support doing more clever things in the future,
-    like doing \"the right thing\" for uncached imports (i.e. exporting
-    environment variables or creating files)
--}
-exprToImport :: Import -> Expr Src X -> StateT (Status IO) IO ()
-exprToImport here expression = do
-    Status {..} <- State.get
-
-    let Import {..} = here
-
-    let ImportHashed {..} = importHashed
-
+            semanticImport <- processImportWithSemanticCache stack import_
+            -- Lens.Family.State.Strict only exports `%=`, not `modifying`...
+            cache %= (Map.insert import_ semanticImport)
+            return semanticImport
+
+-- Checks the "semantic" disk cache; defers to
+-- `processImportWithSemisemanticCache` in general.
+processImportWithSemanticCache
+  :: ImportStack -> ResolvedImport -> StateT Status IO SemanticImport
+-- Pass on non-frozen imports.
+processImportWithSemanticCache stack import_@(ResolvedImport Nothing _ _) =
+    processImportWithSemisemanticCache stack import_
+
+processImportWithSemanticCache stack
+  import_@(ResolvedImport (Just semanticHash) _ _) = do
+    mCached <- lift $ fetchFromSemanticCache semanticHash
+    case mCached of
+        Just bytesStrict -> do
+            let actualHash = Crypto.Hash.hash bytesStrict
+            if semanticHash == actualHash
+                then return ()
+                else throwMissingImport (Imported stack (HashMismatch {expectedHash = semanticHash, ..}))
+
+            let bytesLazy = Data.ByteString.Lazy.fromStrict bytesStrict
+            term <- Dhall.Core.throws (Codec.Serialise.deserialiseOrFail bytesLazy)
+            normalisedImport <- Dhall.Core.throws (Dhall.Binary.decodeExpression term)
+            -- a frozen import does not have any dependencies if loaded from the semantic cache.
+            let graph = Tree.Node import_ []
+            -- TODO: cached imports should have type Expr Src X!
+            return (SemanticImport {normalisedImport = fmap undefined normalisedImport, ..})
+
+        Nothing -> do  -- continue with semi-semantic cache
+            SemanticImport { normalisedImport = normalisedImport, graph = Tree.Node _ imports }
+                <- processImportWithSemisemanticCache stack import_
+
+            -- The semi-semantic cache always returns a semantic hash computed
```

Comment:
Are you sure that you need to check against old standard versions?  The only reason that the semantic hash checks older versions is to support caches generated by older versions of the interpreter.  However, for the semi-semantic cache we don't have to support legacy versions since this is the first version of the interpreter to use the semi-semantic cache.

Reply:
This bit of code concerns the semantic cache! If it's not already in there we load an import via the semi-semantic cache (or ultimately from scratch) and then have to check if it has the expected semantic hash.